tests/test_model.py

```diff
@@ -598,3 +602,90 @@ def test_last_modified(self):
         self.assertEqual(mb.excluded, 'J. Smith')
         self.assertGreaterEqual(mb.last_modified, before_modified_dt)
         self.assertLessEqual(mb.last_modified, after_modified_dt)
+
+
+class TestTransaction(TransactionTestCase):
+    """Unit tests for Transaction class"""
+
+    def test_unique_transactions(self):
+        """Check whether transactions are unique"""
+
+        with self.assertRaisesRegex(IntegrityError, DUPLICATE_CHECK_ERROR):
+            timestamp = datetime_utcnow()
+            Transaction.objects.create(tuid='12345abcd',
+                                       name='test', created_at=timestamp)
+            Transaction.objects.create(tuid='12345abcd',
+                                       name='test', created_at=timestamp)
+
+    def test_created_at(self):
+        """Check creation date is only set when the object is created"""
+
+        before_dt = datetime_utcnow()
+        trx = Transaction.objects.create(tuid='12345abcd',
+                                         name='test', created_at=datetime_utcnow())
+        after_dt = datetime_utcnow()
+
+        self.assertGreaterEqual(trx.created_at, before_dt)
+        self.assertLessEqual(trx.created_at, after_dt)
+
+        trx.save()
+
+        # Check if creation date does not change after saving the object
+        self.assertGreaterEqual(trx.created_at, before_dt)
+        self.assertLessEqual(trx.created_at, after_dt)
+
+
+class TestOperation(TransactionTestCase):
```

Comment:
I'm having difficulties testing this case, because this field is a `CharField` from Django models, whose `choices` options are the values from the enum class `OpType`.

When creating a `Operation` object directly using the call from Django models, it does not raise any error when the field `op_type` receives a explicit OpType value like `ADD` (string) or another string which is not in OpType (for instance, `test`). Any idea to get this under control?

Testing this case in `test_log.py` was not a problem, because the `log_operation` method now controls when `op_type` field is not an instance of `Operation.OpType`.

Reply:
It seems this check would be difficult to perform. After checking [the doc about Django choices](https://docs.djangoproject.com/en/2.2/ref/models/fields/#choices) for `CharField`, it says that "If choices are given, they’re enforced by model validation". However, Django does not check this implicitly.

It does not check if the field's value is one of the choices or even if it is an integer instead of a string, but I added a test checking when `op_type` is `None` as this case is contemplated.

One of the alternatives would  be to declare a `clean()` method validating the field, but it would have to be called every time the object is created. 

As we are already checking this case when the `Operation` object is created in the `log_operation` method from `TransactionsLog`, we decided to decline this test.